File: libredox/src/fs/file.rs
use core::usize;

use io::{Read, Write, Seek, SeekFrom};
use string::{String, ToString};
use vec::Vec;

use syscall::{sys_open, sys_dup, sys_close, sys_execve, sys_fpath, sys_read, sys_write, sys_lseek, sys_fsync};

/// A Unix-style file
pub struct File {
    /// The id for the file
    fd: usize,
}

impl File {
    pub fn exec(path: &str) -> bool {
        unsafe {
            sys_execve((path.to_string() + "\0").as_ptr()) == 0
        }
    }

    /// Open a new file using a path
    pub fn open(path: &str) -> Option<File> {
        unsafe {
            let fd = sys_open((path.to_string() + "\0").as_ptr(), 0, 0);
            if fd == usize::MAX {
                None
            }else{
                Some(File {
                    fd: fd
                })
            }
        }
    }

    /// Duplicate the file
    pub fn dup(&self) -> Option<File> {
        unsafe{
            let new_fd = sys_dup(self.fd);
            if new_fd == usize::MAX {
                None
            } else {
                Some(File {
                    fd: new_fd
                })
            }
        }
    }

    /// Get the canonical path of the file
    pub fn path(&self) -> Option<String> {
        unsafe {
            let mut buf: [u8; 4096] = [0; 4096];
            let count = sys_fpath(self.fd, buf.as_mut_ptr(), buf.len());
            if count == usize::MAX {
                None
            } else {
                Some(String::from_utf8_unchecked(Vec::from(&buf[0..count])))
            }
        }
    }

    /// Flush the io
    pub fn sync(&mut self) -> bool {
        unsafe { sys_fsync(self.fd) == 0 }
    }
}

impl Read for File {
    fn read(&mut self, buf: &mut [u8]) -> Option<usize> {
        unsafe {
            let count = sys_read(self.fd, buf.as_mut_ptr(), buf.len());
            if count == usize::MAX {
                None
            } else {
                Some(count)
            }
        }
    }
}

impl Write for File {
    fn write(&mut self, buf: &[u8]) -> Option<usize> {
        unsafe {
            let count = sys_write(self.fd, buf.as_ptr(), buf.len());
            if count == usize::MAX {
                None
            } else {
                Some(count)
            }
        }
    }
}

impl Seek for File {
    /// Seek a given position
    fn seek(&mut self, pos: SeekFrom) -> Option<usize> {
        let (whence, offset) = match pos {
            SeekFrom::Start(offset) => (0, offset as isize),
            SeekFrom::Current(offset) => (1, offset),
            SeekFrom::End(offset) => (2, offset),
        };

        let position = unsafe { sys_lseek(self.fd, offset, whence) };
        if position == usize::MAX {
            None
        } else {
            Some(position)
        }
    }
}


impl Drop for File {
    fn drop(&mut self) {
        unsafe {
            sys_close(self.fd);
        }
    }
}

File: libc/newlib-redox/newlib/libc/sys/redox/file.c
#include "common.h"

int close(int file){
    return (int)syscall(SYS_CLOSE, (uint)file, 0, 0);
}

int dup(int file){
    return (int)syscall(SYS_DUP, (uint)file, 0, 0);
}

int fpath(int file, char * buf, int len) {
    return (int)syscall(SYS_FPATH, (uint)buf, (uint)len, 0);
}

int fsync(int file) {
    return (int)syscall(SYS_FSYNC, (uint)file, 0, 0);
}

int ftruncate(int file, off_t len){
    return (int)syscall(SYS_FTRUNCATE, (uint)file, (uint)len, 0);
}

int lseek(int file, int ptr, int dir) {
    return (int)syscall(SYS_LSEEK, (uint)file, (uint)ptr, (uint)dir);
}

int link(const char *old, const char *new) {
    return (int)syscall(SYS_LINK, (uint)old, (uint)new, 0);
}

int open(const char *file, int flags, ...) {
    return (int)syscall(SYS_OPEN, (uint)file, (uint)flags, 0);
}

int read(int file, char *ptr, int len) {
    return (int)syscall(SYS_READ, (uint)file, (uint)ptr, (uint)len);
}

int unlink(const char *name) {
    return (int)syscall(SYS_UNLINK, (uint)name, 0, 0);
}

int write(int file, const char *ptr, int len) {
    return (int)syscall(SYS_WRITE, (uint)file, (uint)ptr, (uint)len);
}
